From 1cd0ea61a7e6553a56e68d4fda0f559ba01d6af2 Mon Sep 17 00:00:00 2001
From: Alison Chaiken <alison_chaiken@mentor.com>
Date: Mon, 20 Jul 2015 14:09:40 -0700
Subject: [PATCH 04/23] gstomxvideodec: define format of GstMeta object to hold
 ionBufInfo

Define a GstIonBufFdMeta object based on the GstMeta API that is introduced in gstreamer-1.0 and create methods to write the ionBufInfo into it.
---
 omx/gstomx_meta.h    | 27 +++++++++++++++++++++
 omx/gstomxvideodec.c | 68 ++++++++++++++++++++++++++++++++++++++++++++++++++++
 2 files changed, 95 insertions(+)

diff --git a/omx/gstomx_meta.h b/omx/gstomx_meta.h
index 0e90cb1..cfcdc7e 100644
--- a/omx/gstomx_meta.h
+++ b/omx/gstomx_meta.h
@@ -5,6 +5,8 @@ G_BEGIN_DECLS
 
 #define GST_CAPS_FEATURE_MEMORY_VENUS_IMAGE "memory:Venus"
 
+#include <gst/gst.h>
+#include <gst/video/gstvideometa.h>
 
 const char *decode_color(OMX_COLOR_FORMATTYPE colorcode);
 
@@ -35,6 +37,31 @@ typedef struct GstOmxPlatformPrivateList {
   GstOmxPlatformPrivateEntry *entryList;
 } GstOmxPlatformPrivateList;
 
+struct _GstIonBufFdMeta {
+  GstMeta meta;
+
+  /** ion buffer file descriptor */
+  unsigned int fd;
+  /** Offset from ion buffer base address */
+  unsigned int offset;
+  /** Size of ion buffer */
+  unsigned int size;
+};
+
+typedef struct _GstIonBufFdMeta GstIonBufFdMeta;
+
+GType ionBufFd_meta_api_get_type (void);
+const GstMetaInfo * ionBufFd_meta_get_info (void);
+GstIonBufFdMeta * gst_buffer_add_ionBufFd_meta (GstBuffer *buffer, unsigned int fd, unsigned int offset, unsigned int size);
+
+#define ION_BUF_FD_META_API_TYPE (ionBufFd_meta_api_get_type())
+#define gst_buffer_get_ionfd_meta(b)		\
+  ((GstIonBufFdMeta*)gst_buffer_get_meta((b),ION_BUF_FD_META_API_TYPE))
+#define ION_BUF_FD_META_INFO (ionBufFd_meta_get_info ())
+
+GstIonBufFdMeta * gst_buffer_add_ionBufFd_meta (GstBuffer * buffer,
+					     unsigned int fd, unsigned int offset, unsigned int size);
+
 G_END_DECLS
 
 #endif /* __GST_OMX_META_H__ */
diff --git a/omx/gstomxvideodec.c b/omx/gstomxvideodec.c
index e3ce488..f04e093 100644
--- a/omx/gstomxvideodec.c
+++ b/omx/gstomxvideodec.c
@@ -88,6 +88,13 @@ static OMX_ERRORTYPE gst_omx_video_dec_allocate_output_buffers (GstOMXVideoDec *
 static OMX_ERRORTYPE gst_omx_video_dec_deallocate_output_buffers (GstOMXVideoDec
     * self);
 
+#if defined (USE_OMX_TARGET_VENUS)
+static gboolean ionBufFd_meta_init (GstMeta * meta, gpointer params, GstBuffer * buffer);
+static gboolean ionBufFd_meta_transform (GstBuffer * transbuf, GstMeta * meta,
+			      GstBuffer * buffer, GQuark type, gpointer data);
+static void ionBufFd_meta_free (GstMeta * meta, GstBuffer * buffer);
+#endif
+
 enum
 {
   PROP_0
@@ -417,6 +424,67 @@ gst_omx_video_dec_change_state (GstElement * element, GstStateChange transition)
   return ret;
 }
 
+#if (defined (USE_OMX_TARGET_VENUS))
+GType ionBufFd_meta_api_get_type (void)
+{
+  static volatile GType type;
+  static const gchar *tags[] = { NULL };
+  if (g_once_init_enter (&type)) {
+    GType _type = gst_meta_api_type_register ("ION_BUF_FD_META_API", tags);
+    g_once_init_leave (&type, _type);
+  }
+  return type;
+}
+
+/* initialize HEAD of list of GstMeta items */
+static gboolean ionBufFd_meta_init (GstMeta * meta, gpointer params, GstBuffer * buffer)
+{
+  GstIonBufFdMeta *newmeta = (GstIonBufFdMeta *) meta;
+  newmeta->fd = 0;  /* dangerous: stdin */
+  newmeta->offset = 0;
+  newmeta->size = 0;
+
+  return TRUE;
+}
+
+static void ionBufFd_meta_free (GstMeta * meta, GstBuffer * buffer)
+{
+  return;
+}
+
+static gboolean ionBufFd_meta_transform (GstBuffer * transbuf, GstMeta * meta,
+GstBuffer * buffer, GQuark type, gpointer data)
+{
+  GstIonBufFdMeta * newmeta = (GstIonBufFdMeta *) meta;
+  gst_buffer_add_ionBufFd_meta (transbuf, newmeta->fd, newmeta->offset, newmeta->size);
+  return TRUE;
+}
+
+const GstMetaInfo * ionBufFd_meta_get_info (void)
+{
+  static const GstMetaInfo *meta_info = NULL;
+
+  if (g_once_init_enter (&meta_info)) {
+    const GstMetaInfo *mi = gst_meta_register (ION_BUF_FD_META_API_TYPE,  "ionBufFdMeta", sizeof (GstIonBufFdMeta), ionBufFd_meta_init, ionBufFd_meta_free, ionBufFd_meta_transform);
+      g_once_init_leave (&meta_info, mi);
+  }
+  return meta_info;
+}
+
+GstIonBufFdMeta * gst_buffer_add_ionBufFd_meta (GstBuffer *buffer, unsigned int fd, unsigned int offset, unsigned int size)
+{
+  GstIonBufFdMeta *meta;
+  g_return_val_if_fail (GST_IS_BUFFER (buffer), NULL);
+
+  meta = (GstIonBufFdMeta *) gst_buffer_add_meta (buffer,  ION_BUF_FD_META_INFO, NULL);
+  meta->fd = fd;
+  meta->offset = offset;
+  meta->size = size;
+
+  return meta;
+}
+#endif
+
 static gboolean
 gst_omx_video_dec_fill_buffer (GstOMXVideoDec * self,
     GstOMXBuffer * inbuf, GstBuffer * outbuf)
-- 
1.9.1

